diff --git a/Makefile.am b/Makefile.am
index c587600..a9e0f71 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -61,7 +61,7 @@ lxclipboard_lxclipboard_SOURCES = \
 
 lxclipboard_lxclipboard_VALAFLAGS = \
     --vapidir=$(srcdir)/vapi \
-    --pkg gtk+-2.0 \
+    --pkg gtk+-3.0 \
     --pkg lxclipboard \
     $(NULL)
 
@@ -90,7 +90,7 @@ lxpolkit_lxpolkit_SOURCES = \
 
 lxpolkit_lxpolkit_VALAFLAGS = \
     --vapidir=$(srcdir)/vapi \
-    --pkg gtk+-2.0 \
+    --pkg gtk+-3.0 \
     --pkg lxpolkit \
     $(NULL)
 
@@ -119,7 +119,7 @@ lxsession_db_lxsession_db_SOURCES = \
 
 lxsession_db_lxsession_db_VALAFLAGS = \
     --vapidir=$(srcdir)/vapi \
-    --pkg gtk+-2.0 \
+    --pkg gtk+-3.0 \
     $(NULL)
 
 lxsession_db_lxsession_db_CFLAGS = \
@@ -179,7 +179,7 @@ lxsession_default_apps_lxsession_default_apps_SOURCES = \
 
 lxsession_default_apps_lxsession_default_apps_VALAFLAGS = \
     --vapidir=$(srcdir)/vapi \
-    --pkg gtk+-2.0 \
+    --pkg gtk+-3.0 \
     --pkg dbus-glib-1 \
     --pkg gio-2.0 \
     --pkg posix \
diff --git a/lxsession-default-apps/combobox.vala b/lxsession-default-apps/combobox.vala
index 32d3107..43b7d9e 100644
--- a/lxsession-default-apps/combobox.vala
+++ b/lxsession-default-apps/combobox.vala
@@ -767,7 +767,7 @@ namespace LDefaultApps
             if (entry_name != null)
             {
                 var entry_default = builder.get_object (entry_name) as Entry;
-                entry_default.hide_all();
+                entry_default.hide();
             }
         }
 
@@ -791,7 +791,7 @@ namespace LDefaultApps
                 }
                 else
                 {
-                    entry.hide_all();
+                    entry.hide();
                 }
             }
         });
diff --git a/lxsession-default-apps/main.vala b/lxsession-default-apps/main.vala
index 3b7f868..8d11df9 100644
--- a/lxsession-default-apps/main.vala
+++ b/lxsession-default-apps/main.vala
@@ -70,7 +70,7 @@ namespace LDefaultApps
             Process.close_pid (pid);
             MainWindows win = new MainWindows(kf);
             win.show_all();
-            this.hide_all();
+            this.hide();
         }
 
         public KeyFile get_keyfile()
@@ -390,13 +390,13 @@ namespace LDefaultApps
             init_list_view(known_apps);
             load_autostart(Environment.get_variable("XDG_CURRENT_DESKTOP"));
             known_apps.set_model (get_autostart_list ());
-            known_apps_box.hide_all();
+            known_apps_box.hide();
 
             disable_autostart_combobox.changed.connect (() => {
                 if (return_combobox_text(disable_autostart_combobox) == "all")
                 {
-                    auto_vbox.hide_all();
-                    running_apps.hide_all();
+                    auto_vbox.hide();
+                    running_apps.hide();
                 }
                 else
                 {
@@ -410,7 +410,7 @@ namespace LDefaultApps
                 }
                 else
                 {
-                    known_apps_box.hide_all();
+                    known_apps_box.hide();
                 }
                 dbus_backend.Set("disable_autostart", "", return_combobox_text(disable_autostart_combobox));
             });
