diff --git a/cups/cups/usersys.c b/cups/cups/usersys.c
index 9b4e07e..ad76b4c 100644
--- a/cups/usersys.c
+++ b/cups/usersys.c
@@ -737,6 +737,49 @@
 #endif /* HAVE_GSSAPI */
 
 
+static int
+_cupsUnixSocketAccess(const char* name)
+{
+	struct sockaddr_un s_un;
+	int sock = -1;
+	int ret = 0;
+
+	bzero(&s_un, sizeof(s_un));
+
+	strncpy(s_un.sun_path, name,sizeof s_un.sun_path);
+	s_un.sun_family = AF_UNIX;
+
+	if (((sock = socket(AF_UNIX, SOCK_STREAM, 0)) > 0) &&
+	    (connect(sock, (struct sockaddr *)&s_un,SUN_LEN(&s_un)) == 0))
+	       ret = 1;
+
+	if (sock >= 0) close(sock);
+	return ret;
+}
+
+static int
+_cupsLocalhostAccess(int port)
+{
+	struct sockaddr_in s_in;
+	int sock = -1;
+	int ret = 0;
+
+	bzero(&s_in, sizeof(s_in));
+
+	s_in.sin_family = AF_INET;
+	s_in.sin_port = htons(port);
+
+	if (inet_aton("127.0.0.1", (struct in_addr*)&(s_in.sin_addr)) == 0)
+	    return 0;
+
+	if (((sock = socket(AF_INET, SOCK_STREAM, 0)) > 0) &&
+	    (connect(sock, (struct sockaddr *)&s_in,sizeof(s_in)) == 0))
+	       ret = 1;
+
+	if (sock >= 0) close(sock);
+	return ret;
+}
+
 /*
  * '_cupsSetDefaults()' - Set the default server, port, and encryption.
  */
