diff --git a/gtk/gtkmenushell.c b/gtk/gtkmenushell.c
index 939d5f4..b02436e 100644
--- a/gtk/gtkmenushell.c
+++ b/gtk/gtkmenushell.c
@@ -2079,6 +2079,7 @@ gtk_menu_shell_tracker_insert_func (GtkMenuTrackerItem *item,
 
       widget = gtk_model_menu_item_new ();
       g_object_bind_property (item, "label", widget, "text", G_BINDING_SYNC_CREATE);
+      g_object_bind_property (item, "icon", widget, "icon", G_BINDING_SYNC_CREATE);
 
       submenu = GTK_MENU_SHELL (gtk_menu_new ());
 
