diff -ur cups-1.7.0.old/scheduler/cups.sh.in cups-1.7.0/scheduler/cups.sh.in
--- cups-1.7.0.old/scheduler/cups.sh.in	2012-10-15 21:02:10.000000000 +0000
+++ cups-1.7.0/scheduler/cups.sh.in	2013-12-17 07:07:56.349654612 +0000
@@ -42,19 +42,6 @@
 		ECHO_ERROR=:
 		;;
 
-	Darwin*)
-		. /etc/rc.common
-
-		if test "${CUPS:=-YES-}" = "-NO-"; then
-			exit 0
-		fi
-
-        	IS_ON=:
-		ECHO=ConsoleMessage
-		ECHO_OK=:
-		ECHO_ERROR=:
-		;;
-
 	Linux*)
 		IS_ON=/bin/true
 		if test -f /etc/init.d/functions; then
@@ -86,7 +73,7 @@
 # found, then Greenwich Mean Time (GMT) will probably be used.
 #
 
-for file in /etc/TIMEZONE /etc/rc.config /etc/sysconfig/clock; do
+for file in /etc/sysconfig/clock; do
 	if test -f $file; then
 	        . $file
 	fi
@@ -160,10 +143,6 @@
 case $1 in
 	start | restart | reload)
 		if $IS_ON cups; then
-			if test -x /sbin/portrelease; then
-				/sbin/portrelease cups
-			fi
-
 			if test "$pid" != ""; then
 				kill -HUP $pid
 			else
