Index: libqpol/src/module_compiler.c
===================================================================
--- libqpol/src/module_compiler.c.orig
+++ libqpol/src/module_compiler.c
@@ -291,6 +291,7 @@ role_datum_t *declare_role(void)
 		assert(0);	       /* should never get here */
 	}
 	}
+	return 0;
 }
 
 type_datum_t *declare_type(unsigned char primary, unsigned char isattr)
@@ -353,6 +354,7 @@ type_datum_t *declare_type(unsigned char
 		assert(0);	       /* should never get here */
 	}
 	}
+	return 0;
 }
 
 static int user_implicit_bounds(hashtab_t users_tab, char *user_id, user_datum_t * user)
@@ -482,6 +484,7 @@ user_datum_t *declare_user(void)
 		assert(0);	       /* should never get here */
 	}
 	}
+	return 0;
 }
 
 /* Return a type_datum_t for the local avrule_decl with the given ID.
@@ -822,6 +825,7 @@ int require_role(int pass)
 		assert(0);	       /* should never get here */
 	}
 	}
+	return 0;
 }
 
 static int require_type_or_attribute(int pass, unsigned char isattr)
@@ -873,6 +877,7 @@ static int require_type_or_attribute(int
 		assert(0);	       /* should never get here */
 	}
 	}
+	return 0;
 }
 
 int require_type(int pass)
@@ -932,6 +937,7 @@ int require_user(int pass)
 		assert(0);	       /* should never get here */
 	}
 	}
+	return 0;
 }
 
 int require_bool(int pass)
@@ -979,6 +985,7 @@ int require_bool(int pass)
 		assert(0);	       /* should never get here */
 	}
 	}
+	return 0;
 }
 
 int require_sens(int pass)
@@ -1041,6 +1048,7 @@ int require_sens(int pass)
 		assert(0);	       /* should never get here */
 	}
 	}
+	return 0;
 }
 
 int require_cat(int pass)
@@ -1093,6 +1101,7 @@ int require_cat(int pass)
 		assert(0);	       /* should never get here */
 	}
 	}
+	return 0;
 }
 
 static int is_scope_in_stack(scope_datum_t * scope, scope_stack_t * stack)
Index: libqpol/src/policy_define.c
===================================================================
--- libqpol/src/policy_define.c.orig
+++ libqpol/src/policy_define.c
@@ -1512,6 +1512,7 @@ int define_compute_type(int which)
 		assert(0);	       /* should never get here */
 	}
 	}
+	return 0;
 }
 
 avrule_t *define_cond_compute_type(int which)
